### ISSUE TITLE
Triangle Counting Validity: FAIL

### ISSUE THREAD

**Issue by wzbxpy** [gunrock/gunrock#680]
**Summary:**
Triangle Counting Validity: FAIL in soc-LiveJournal dataset

**Reproducibility:**
Always

**To Reproduce:**
make the tc
run
./tc market ../../dataset/large/soc-LiveJournal1/soc-LiveJournal1.mtx --quick --device=0 --sort-csr --64bit-SizeT=false,true --64bit-VertexT=false,true

**Actual Results:**

Loading Matrix-market coordinate-formatted graph ...
  Reading meta data from ../../dataset/large/soc-LiveJournal1/soc-LiveJournal1.mtx.meta
  Reading edge lists from ../../dataset/large/soc-LiveJournal1/soc-LiveJournal1.mtx.coo_edge_pairs
Reading from ../../dataset/large/soc-LiveJournal1/soc-LiveJournal1.mtx.coo_edge_pairs, typeId = 262, targetId = 262, length = 68993773
  Assigning 1 to all 68993773 edges
  Substracting 1 from node Ids...
  Removed 518382 duplicate edges and self circles.
  graph loaded as COO in 5.781388s.
Converting 4847571 vertices, 68475391 directed edges ( ordered tuples) to CSR format...Done (1s).
Degree Histogram (4847571 vertices, 68475391 edges):
    Degree 0: 553239 (11.412705 %)
    Degree 2^0: 941799 (19.428266 %)
    Degree 2^1: 797094 (16.443163 %)
    Degree 2^2: 749096 (15.453018 %)
    Degree 2^3: 680822 (14.044601 %)
    Degree 2^4: 565267 (11.660830 %)
    Degree 2^5: 354208 (7.306917 %)
    Degree 2^6: 146734 (3.026959 %)
    Degree 2^7: 45005 (0.928403 %)
    Degree 2^8: 11773 (0.242864 %)
    Degree 2^9: 2474 (0.051036 %)
    Degree 2^10: 44 (0.000908 %)
    Degree 2^11: 7 (0.000144 %)
    Degree 2^12: 5 (0.000103 %)
    Degree 2^13: 3 (0.000062 %)
    Degree 2^14: 1 (0.000021 %)

==============================================
64bit-VertexT=false 64bit-SizeT=false 64bit-ValueT=true undirected=false advance-mode=LB
Using advance mode LB
Using filter mode CULL
__________________________
--------------------------
Run 0 elapsed: 119.541883 ms, #iterations = 1
Triangle Counting Validity: FAIL: [0]: 61 != 0
result[...61, 564, 244, 167, 193, ...]
reference[...0, 0, 0, 0, 0, ...]
2959121 errors occurred.
[tc] finished.
 avg. elapsed: 119.541883 ms
 iterations: 1
 min. elapsed: 119.541883 ms
 max. elapsed: 119.541883 ms
 load time: 7909.12 ms
 preprocess time: 5201.000000 ms
 postprocess time: 60.343027 ms
 total time: 5862.132072 ms

**System Configuration:**

GPU: V100
OS: Ubuntu 18.04.1 LTS
CUDA Version: 10.2
gcc version 7.4.0

**Comment by Laurawly** [comment]
@wzbxpy it looks from the output that the reference is wrong instead of the Gunrock results . I’ll double check it. Also the tc in Gunrock is based on this paper: https://dl.acm.org/doi/10.1145/2915516.2915521 and the method is intersection based method.

**Comment by wzbxpy** [comment]
> 
> 
> @wzbxpy it looks from the output that the reference is wrong instead of the Gunrock results . I’ll double check it. Also the tc in Gunrock is based on this paper: https://dl.acm.org/doi/10.1145/2915516.2915521 and the method is intersection based method.

So how to use the triangle counting in Fast Parallel BFS-Based Triangle Counting on GPUs - Leyuan Wang, John D Owens (Univeristy of California, Davis), have you publish the code of it?

**Comment by Laurawly** [comment]
@wzbxpy we haven't published the code for it yet. We are working on cleaning the code.

**Comment by Laurawly** [comment]
For the reference code bug, I confirmed that it's not working any more with any dataset. But before this commit (b4717f22106bccdeab1df88af583a90ff6f57ec7), the reference code was tested and worked properly. @neoblizz Do you recall testing it under experiments/tc after the commit?

**Comment by wzbxpy** [comment]
> @wzbxpy we haven't published the code for it yet. We are working on cleaning the code.

Can you share the code before cleaning to me?

**Comment by jowens** [comment]
Our focus is on getting triangle counting and subgraph matching into Gunrock. That is more important to us than cleaning up, or releasing, the code we used in this summer's paper.

**Comment by Laurawly** [comment]
@wzbxpy yes, we are working on getting the code to Gunrock now. I’ll ping you once it gets merged

**Comment by wzbxpy** [comment]
OK, thanks
